File: src/features/typing/data/ts/permutations.ts
function permute(nums: number[]): number[][] {
  const result: number[][] = [];
  const n = nums.length;
  const backtrack = (start: number): void => {
    if (start === n) {
      result.push([...nums]);
      return;
    }
    for (let i = start; i < n; i++) {
      [nums[start], nums[i]] = [nums[i], nums[start]];
      backtrack(start + 1);
      [nums[start], nums[i]] = [nums[i], nums[start]];
    }
  };
  backtrack(0);
  return result;
}

File: src/features/typing/data/dart/permutations.dart
class Solution {
  List<List<int>> permute(List<int> nums) {
    final result = <List<int>>[];
    final n = nums.length;
    void backtrack(int start) {
      if (start == n) {
        result.add(List.of(nums));
        return;
      }
      for (int i = start; i < n; i++) {
        _swap(nums, start, i);
        backtrack(start + 1);
        _swap(nums, start, i);
      }
    }
    backtrack(0);
    return result;
  }

  void _swap(List<int> nums, int i, int j) {
    final tmp = nums[i];
    nums[i] = nums[j];
    nums[j] = tmp;
  }
}
